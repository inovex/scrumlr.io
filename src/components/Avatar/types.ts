export const AVATAR_SKIN_COLORS = ["Tanned", "Yellow", "Pale", "Light", "Brown", "DarkBrown", "Black"] as const;
export type AvatarSkinColor = (typeof AVATAR_SKIN_COLORS)[number];

export const AVATAR_TOP_TYPES = [
  "Eyepatch",
  "Hat",
  "Hijab",
  "LongHairBigHair",
  "LongHairBob",
  "LongHairBun",
  "LongHairCurly",
  "LongHairCurvy",
  "LongHairDreads",
  "LongHairFrida",
  "LongHairFro",
  "LongHairFroBand",
  "LongHairMiaWallace",
  "LongHairNotTooLong",
  "LongHairShavedSides",
  "LongHairStraight",
  "LongHairStraight2",
  "LongHairStraightStrand",
  "NoHair",
  "ShortHairDreads01",
  "ShortHairDreads02",
  "ShortHairFrizzle",
  "ShortHairShaggyMullet",
  "ShortHairShortCurly",
  "ShortHairShortFlat",
  "ShortHairShortRound",
  "ShortHairShortWaved",
  "ShortHairSides",
  "ShortHairTheCaesar",
  "ShortHairTheCaesarSidePart",
  "Turban",
  "WinterHat1",
  "WinterHat2",
  "WinterHat3",
  "WinterHat4",
] as const;
export type AvatarTopType = (typeof AVATAR_TOP_TYPES)[number];

export const AVATAR_CLOTHE_COLORS = [
  "Black",
  "Blue01",
  "Blue02",
  "Blue03",
  "Gray01",
  "Gray02",
  "Heather",
  "PastelBlue",
  "PastelGreen",
  "PastelOrange",
  "PastelRed",
  "PastelYellow",
  "Pink",
  "Red",
  "White",
] as const;
export type AvatarClotheColor = (typeof AVATAR_CLOTHE_COLORS)[number];

export const AVATAR_GRAPHIC_TYPES = ["Bat", "Cumbia", "Deer", "Diamond", "Hola", "Pizza", "Resist", "Selena", "Bear", "SkullOutline", "Skull"] as const;
export type AvatarGraphicType = (typeof AVATAR_GRAPHIC_TYPES)[number];

export const AVATAR_CLOTHE_TYPES = ["BlazerShirt", "BlazerSweater", "CollarSweater", "GraphicShirt", "Hoodie", "Overall", "ShirtCrewNeck", "ShirtScoopNeck", "ShirtVNeck"] as const;
export type AvatarClotheType = (typeof AVATAR_CLOTHE_TYPES)[number];

export const AVATAR_HAIR_COLORS = ["Auburn", "Black", "Blonde", "BlondeGolden", "Brown", "BrownDark", "PastelPink", "Blue", "Platinum", "Red", "SilverGray"] as const;
export type AvatarHairColor = (typeof AVATAR_HAIR_COLORS)[number];

export const AVATAR_FACIAL_HAIR_COLORS = ["Auburn", "Black", "Blonde", "BlondeGolden", "Brown", "BrownDark", "Platinum", "Red"] as const;
export type AvatarFacialHairColor = (typeof AVATAR_FACIAL_HAIR_COLORS)[number];

export const AVATAR_FACIAL_HAIR_TYPES = ["Blank", "BeardMedium", "BeardLight", "BeardMajestic", "MoustacheFancy", "MoustacheMagnum"] as const;
export type AvatarFacialHairType = (typeof AVATAR_FACIAL_HAIR_TYPES)[number];

export const AVATAR_ACCESSORIES_TYPES = ["Blank", "Kurt", "Prescription01", "Prescription02", "Round", "Sunglasses", "Wayfarers"] as const;
export type AvatarAccessoriesType = (typeof AVATAR_ACCESSORIES_TYPES)[number];

export const AVATAR_EYE_TYPES = ["Close", "Cry", "Default", "Dizzy", "EyeRoll", "Happy", "Hearts", "Side", "Squint", "Surprised", "Wink", "WinkWacky"] as const;
export type AvatarEyeType = (typeof AVATAR_EYE_TYPES)[number];

export const AVATAR_EYEBROW_TYPES = [
  "Angry",
  "AngryNatural",
  "Default",
  "DefaultNatural",
  "FlatNatural",
  "FrownNatural",
  "RaisedExcited",
  "RaisedExcitedNatural",
  "SadConcerned",
  "SadConcernedNatural",
  "UnibrowNatural",
  "UpDown",
  "UpDownNatural",
] as const;
export type AvatarEyebrowType = (typeof AVATAR_EYEBROW_TYPES)[number];

export const AVATAR_MOUTH_TYPES = ["Concerned", "Default", "Disbelief", "Eating", "Grimace", "Sad", "ScreamOpen", "Serious", "Smile", "Tongue", "Twinkle", "Vomit"] as const;
export type AvatarMouthType = (typeof AVATAR_MOUTH_TYPES)[number];
