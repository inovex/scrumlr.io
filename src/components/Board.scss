@import "common";

.board {
  --n: 1;

  display:grid;
  grid-auto-flow: column;
  grid-auto-columns: calc(100%/min(var(--real-columns), var(--n)));
  overflow-x:auto;
  scroll-snap-type: x mandatory;

  height: 100vh;
}

@for $i from 1 through 12 {
  @media screen and (min-width: $i*$min-column-width) {
    .board {
      --n: #{$i};
    }
  }
}