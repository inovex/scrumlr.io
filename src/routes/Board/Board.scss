.board-page {
  outline: none;
  display: flex;
  flex-direction: column;
  min-height: 430px;
  overflow: hidden;
}

.board-page__column-view {
  flex: 1;
}

.phase-splash__animation-enter, .phase-splash__animation-appear {
  // FIXME adding scale on enter breaks carousel calculation
  //transform: scale(.1);
  opacity: .005 !important;
}

.phase-splash__animation-appear.phase-splash__animation-appear-active,
.phase-splash__animation-enter.phase-splash__animation-enter-active {
  transform: scale(1);
  opacity: .95 !important;
  transition: opacity 200ms ease-in, transform 300ms ease-in;
}

.phase-splash__animation-leave {
  transform: scale(1);
  opacity: .95 !important;
}

.phase-splash__animation-leave.phase-splash__animation-leave-active {
  transform: scale(4);
  opacity: .005 !important;
  transition: opacity 300ms ease-in, transform 300ms ease-in;
}
