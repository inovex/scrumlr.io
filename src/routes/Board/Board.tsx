import {useSelector} from "react-redux";
import LoadingScreen from "components/LoadingScreen/LoadingScreen";
import BoardComponent from "components/Board/Board";
import Column from "components/Column/Column";
import {ApplicationState} from "types/store";
import Parse from "parse";
import Note from "components/Note/Note";
import JoinRequest from "components/JoinRequest/JoinRequest";

function Board() {
  const state = useSelector((applicationState: ApplicationState) => ({
    board: applicationState.board,
    notes: applicationState.notes,
    joinRequests: applicationState.joinRequests,
    users: applicationState.users,
  }));

  let joinRequestComponent;
  if (state.users.admins.find((user) => user.id === Parse.User.current()!.id) !== undefined) {
    const pendingJoinRequests = state.joinRequests.filter((joinRequest) => joinRequest.status === "pending");
    if (pendingJoinRequests && pendingJoinRequests.length > 0) {
      joinRequestComponent = <JoinRequest joinRequests={pendingJoinRequests} />;
    }
  }
  let boardstatus = "Public Session";
  const joinConfirmationRequired = state.board.data?.joinConfirmationRequired;
  const accessCode = state.board.data?.accessCode;
  if (joinConfirmationRequired === true || (accessCode !== undefined && accessCode !== "")) {
    boardstatus = "Private Session";
  }

  if (state.board.status === "pending") {
    return <LoadingScreen />;
  }
  if (state.board.status === "ready") {
    return (
      <>
        {joinRequestComponent}
        <BoardComponent name={state.board.data!.name} boardstatus={boardstatus}>
          {state.board.data!.columns.map((column) => (
            <Column key={column.id} id={column.id!} name={column.name} color={column.color}>
              {state.notes
                .filter((note) => note.columnId === column.id)
                .map((note) => (
                  <Note key={note.id} text={note.text} authorId={note.author} />
                ))}
            </Column>
          ))}
        </BoardComponent>
      </>
    );
  }
  return <LoadingScreen />;
}
export default Board;
