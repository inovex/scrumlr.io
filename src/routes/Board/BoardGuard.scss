@import "constants/style";

.board-guard {
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
}

.board-guard__denied-link {
  color: $color-planning-pink;
  transition: all 0.08s ease-out;

  &:hover {
    filter: brightness(1.4);
    transform: scale(1.02);
  }
}

.board-guard__info {
  font-size: $text-size--large;
}

[theme="dark"] {
  .board-guard {
    background-color: $color-dark-one;
  }

  .board-guard__info {
    color: $color-white;
  }
}
