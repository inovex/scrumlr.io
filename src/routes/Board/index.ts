export * from "./Board";
export * from "./BoardGuard";
