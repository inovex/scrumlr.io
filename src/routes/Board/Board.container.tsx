import { AnyAction, Dispatch } from 'redux';
import { isLoaded, getVal } from 'react-redux-firebase';
import * as Raven from 'raven-js';
import { debounce, difference } from 'lodash';

import { SETUP_COMPLETED } from '../../actions';
import { BoardProps } from './Board';
import {
  BoardConfig,
  FirebaseProp,
  StoreState,
  BoardCards,
  Card,
  Optional,
  PrivateBoardData,
  PublicBoardData
} from '../../types';
import { authController } from '../../controller/auth';
import { Crypto } from '../../util/crypto';

export const mapStateToProps = (
  state: StoreState,
  ownProps: BoardProps
): BoardProps => {
  const { app, fbState } = state;
  const firebase = ownProps.firebase as FirebaseProp;

  const boardSelector = `boards/${ownProps.match.params.id}/private`;
  const boardPrintUrl = `/print/${ownProps.match.params.id}`;
  const auth: firebase.User | any = firebase.auth().currentUser || {};

  const board: Optional<PrivateBoardData> = getVal(
    fbState,
    `data/${boardSelector}`,
    undefined
  );

  if (!isLoaded(board) || !board) {
    return { boardSelector, boardConfig: undefined, auth } as any;
  }

  const cards: BoardCards = board.cards || {};
  const boardConfig: BoardConfig = board.config;

  const users = isLoaded(board) ? board.users : {};

  const adminUsers = boardConfig.adminUsers || {};

  //There should be only one board admin check - here!
  const isBoardAdmin = isLoaded(board)
    ? auth.uid === boardConfig.creatorUid || adminUsers[auth.uid]
    : false;

  let focusedCard: Optional<Card> = undefined;
  if (boardConfig.focusedCardId) {
    focusedCard = cards[boardConfig.focusedCardId];
  }

  let timerExpiration: Optional<string> = undefined;
  if (boardConfig.timerExpiration) {
    timerExpiration = boardConfig.timerExpiration;
  }

  const publicBoardSelector = `boards/${ownProps.match.params.id}/public`;
  const publicBoard: Optional<PublicBoardData> = getVal(
    fbState,
    `data/${publicBoardSelector}`,
    undefined
  );

  let waitingUsers: {
    uid: string;
    name: string;
    image: string;
    time: Date;
  }[] = [];
  if (isLoaded(publicBoard) && publicBoard) {
    const { accessAuthorized, applicants } = publicBoard;

    if (applicants) {
      const waitingUserUids = difference(
        Object.keys(applicants),
        Object.keys(accessAuthorized || {})
      );
      waitingUsers = waitingUserUids
        .map(uid => ({
          uid,
          ...applicants[uid],
          time: new Date(applicants[uid].time)
        }))
        .sort((a, b) => {
          if (a.time > b.time) {
            return 1;
          } else if (a.time < b.time) {
            return -1;
          }
          return 0;
        });
    }
  }

  function acceptUser(uid: string, accept: boolean) {
    firebase
      .set(`${publicBoardSelector}/accessAuthorized/${uid}`, accept)
      .catch((err: any) => {
        Raven.captureMessage('unable to accept user access', {
          extra: {
            reason: err.message,
            uid: auth.uid,
            boardId: boardSelector,
            user: uid,
            accept
          }
        });
      });
  }

  // TODO cleanup symmetric key sync
  if (isBoardAdmin && users) {
    const keyStore = board.keyStore;

    if (keyStore) {
      const crypto = new Crypto();
      crypto.initKeypair().then(() => {
        const key = keyStore[crypto.getPublicKey()!!];

        if (key) {
          crypto.importSymmetricKey(key).then(() => {
            for (let uid in users) {
              if (keyStore && !keyStore[users[uid].publicKey!!]) {
                crypto
                  .exportSymmetricKey(users[uid].publicKey!!)
                  .then(exportedKey => {
                    firebase.set(
                      `${boardSelector}/keyStore/${users[uid].publicKey}`,
                      exportedKey
                    );
                  });
              }
            }
          });
        }
      });
    }
  }

  function onToggleShowAuthor() {
    firebase
      .update(`${boardSelector}/config`, {
        showAuthor: !boardConfig.showAuthor
      })
      .catch((err: Error) => {
        Raven.captureMessage('Could not toggle show author state', {
          extra: { reason: err.message, uid: auth.uid, boardId: boardSelector }
        });
      });
  }

  function onToggleReadyState() {
    firebase
      .update(`${boardSelector}/users/${auth.uid}`, {
        ready: !board!!.users[auth.uid].ready
      })
      .catch((err: Error) => {
        Raven.captureMessage('Could not toggle user state', {
          extra: { reason: err.message, uid: auth.uid, boardId: boardSelector }
        });
      });
  }

  function onFocusCard(cardId: string) {
    const { focusedCardId } = boardConfig;
    firebase
      .set(
        `${boardSelector}/config/focusedCardId`,
        focusedCardId !== cardId ? cardId : null
      )
      .catch((err: any) => {
        Raven.captureMessage('Could not focus card', {
          extra: {
            reason: err.message,
            uid: auth.uid,
            boardId: boardSelector,
            focusedCardId
          }
        });
      });
  }

  const username = auth.displayName || undefined;
  const email = auth.email || undefined;
  const isAnonymous = auth.isAnonymous;

  const onChangeUsername = debounce((username: string) => {
    const user = firebase.auth().currentUser;
    if (user) {
      user
        .updateProfile({
          displayName: username,
          photoURL: user.photoURL
        })
        .then(() => {
          firebase.update(`${boardSelector}/users/${auth.uid}`, {
            name: username
          });
        });
    }
  }, 2000);

  const onChangeBoardName = debounce((boardName: string) => {
    const { focusedCardId } = boardConfig;
    firebase
      .set(`${boardSelector}/config/name`, boardName)
      .catch((err: any) => {
        Raven.captureMessage('Could not set boardname', {
          extra: {
            reason: err.message,
            uid: auth.uid,
            boardId: boardSelector,
            focusedCardId
          }
        });
      });
  }, 2000);

  const onDeleteTimer = () => {
    firebase
      .set(`${boardSelector}/config/timerExpiration`, null)
      .catch((err: any) => {
        Raven.captureMessage('Unable to delete timer', {
          extra: {
            reason: err.message,
            uid: auth.uid,
            boardId: boardSelector
          }
        });
      });
  };

  function onSwitchPhaseIndex(delta: number) {
    firebase
      .ref(`${boardSelector}/config/guidedPhase`)
      .transaction((phase: number) => phase + delta)
      .catch((err: Error) => {
        Raven.captureMessage('Could not switch phase index', {
          extra: { reason: err.message, uid: auth.uid, boardId: boardSelector }
        });
      });

    const updateUsers = {};
    Object.keys(board!!.users).forEach(uid => {
      updateUsers[`${uid}/ready`] = false;
    });

    console.log(updateUsers);
    firebase
      .ref(`${boardSelector}/users`)
      .update(updateUsers)
      .catch((err: Error) => {
        Raven.captureMessage('Could reset ready status for users', {
          extra: { reason: err.message, boardId: boardSelector }
        });
      });

    // Always remove card selection if new phase is entered.
    if (boardConfig.focusedCardId) {
      onFocusCard(boardConfig.focusedCardId);
    }
  }

  return {
    ...ownProps,
    cards,
    boardSelector,
    boardConfig,
    boardPrintUrl,
    users,
    focusedCard,
    isBoardAdmin,
    username,
    email,
    timerExpiration,
    isAnonymous,
    uid: auth.uid,
    onToggleReadyState,
    onFocusCard,
    onSwitchPhaseIndex,
    isShowAuthor: Boolean(boardConfig.showAuthor),
    onSignOut: authController(firebase).signOut,
    onChangeBoardName,
    onChangeUsername,
    onDeleteTimer,
    onToggleShowAuthor,
    onRegisterCurrentUser: () => null, // will be filled in mergeProps
    waitingUsers,
    acceptUser,
    ...app,

    // other props, only used in mergeProps
    auth
  };
};

export function mapDispatchToProps(dispatch: Dispatch<AnyAction>) {
  return {
    dispatch
  };
}

export function mergeProps(
  stateProps: BoardProps,
  { dispatch }: { dispatch: Dispatch<any> },
  ownProps: BoardProps
): BoardProps {
  function onRegisterCurrentUser() {
    dispatch({ type: SETUP_COMPLETED });
  }

  return {
    ...ownProps,
    ...stateProps,
    onRegisterCurrentUser
  };
}
