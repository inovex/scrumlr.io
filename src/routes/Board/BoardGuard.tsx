import {useEffect} from "react";
import store from "store";
import {ActionFactory} from "store/action";
import {RouteComponentProps} from "react-router";
import {ApplicationState} from "types/store";
import {useSelector} from "react-redux";
import LoadingIndicator from "components/LoadingIndicator/LoadingIndicator";
import Board from "./Board";
import "./BoardGuard.scss";

export type BoardGuardProps = RouteComponentProps<{id: string}>;

const BoardGuard = (props: BoardGuardProps) => {
  const boardStatus = useSelector((state: ApplicationState) => state.board.status);

  useEffect(() => {
    const boardId = props.match.params.id;
    store.dispatch(ActionFactory.joinBoard(boardId));

    return () => {
      store.dispatch(ActionFactory.leaveBoard());
    };
  }, [props.match.params.id]);

  if (boardStatus === "accepted" || boardStatus === "ready") {
    return <Board />;
  }
  if (boardStatus === "rejected") {
    return (
      <div className="board-guard">
        <p className="board-guard__info">You have been rejected.</p>
        <a href="/" className="board-guard__denied-link">
          Return to homepage
        </a>
      </div>
    );
  }
  return (
    <div className="board-guard">
      <LoadingIndicator />
      <p className="board-guard__info">Waiting for approval.</p>
    </div>
  );
};

export default BoardGuard;
