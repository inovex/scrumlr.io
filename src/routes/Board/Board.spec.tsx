import * as React from 'react';
import { mount, shallow, ShallowWrapper } from 'enzyme';

import { mockBoardConfig, mockCard, mockUser } from '../../builder';
import { BoardCards, BoardUsers } from '../../types';
import { Board, BoardProps } from './Board';
import LoadingScreen from '../../components/LoadingScreen/LoadingScreen';

jest.mock(`!svg-inline-loader!./logo.svg`, () => 'svg', {
  virtual: true
});
jest.mock(`!svg-inline-loader!./logo-s.svg`, () => 'svg', {
  virtual: true
});

describe('<Board />', () => {
  let shallowWrapper: ShallowWrapper<BoardProps, {}>;
  let mockProps: BoardProps;

  beforeEach(() => {
    const cards = {};
    const boardConfig = mockBoardConfig();
    const users = {};
    mockProps = {
      cards,
      boardConfig,
      users,
      boardUrl: '/boards/-foobar',
      isBoardAdmin: false,
      uid: '',
      registered: false,
      setupCompleted: false,

      onRegisterCurrentUser: jest.fn(),
      onToggleReadyState: jest.fn(),
      onFocusCard: jest.fn(),
      onSwitchPhaseIndex: jest.fn(),
      onSignOut: jest.fn(),
      onChangeBoardName: jest.fn(),

      match: {
        params: { id: '-foobar' },
        isExact: true,
        path: '/boards/:id',
        url: '/boards/-foobar'
      },
      location: null as any,
      history: null as any,
      boardSelector: 'test',
      boardPrintUrl: '/print',
      onChangeUsername: jest.fn(),
      onChangeEmail: jest.fn(),
      isAnonymous: false
    };
  });

  describe('lifecycle', () => {
    it('should register current user once component has been mounted', () => {
      const spy = jest.fn();
      mount(<Board {...mockProps} onRegisterCurrentUser={spy} />);
      expect(spy).toHaveBeenCalled();
    });
  });

  describe('getUsername instance method', () => {
    it('should return correct username from given userlist in getUsername method', () => {
      const users: BoardUsers = {
        user1: mockUser()
      };
      shallowWrapper = shallow(<Board {...mockProps} users={users} />);
      expect((shallowWrapper.instance() as any).getUsername('user1')).toEqual(
        users.user1.name
      );
    });

    it('should return "Unknown user" if getUsername is called with unknown user id', () => {
      const users: BoardUsers = {
        user1: mockUser()
      };
      shallowWrapper = shallow(<Board {...mockProps} users={users} />);
      expect((shallowWrapper.instance() as any).getUsername('user2')).toEqual(
        undefined
      );
    });
  });

  describe('export', () => {
    let cards: BoardCards;
    let users: BoardUsers;

    beforeEach(() => {
      users = {
        user1: mockUser({ name: 'User1' }),
        user2: mockUser({ name: 'User2' })
      };
      cards = {
        card1: mockCard({
          text: 'Text card 1',
          authorUid: 'user1',
          userVotes: { user1: 10, user2: 3 },
          type: 'positive'
        }),
        card2: mockCard({
          text: 'Text card 2',
          authorUid: 'user2',
          userVotes: { user1: 3, user2: -4 },
          type: 'negative'
        }),
        card3: mockCard({
          text: 'Text card 3',
          authorUid: 'user1',
          userVotes: { user2: -4 },
          parent: 'card1',
          type: 'positive'
        })
      };
    });

    it('should be possible to generate CSV from cards', () => {
      const separatorChar = '|';
      shallowWrapper = shallow(
        <Board {...mockProps} cards={cards} users={users} />
      );
      expect(
        (shallowWrapper.instance() as any).exportCsv(separatorChar)
      ).toMatchSnapshot();
    });

    it('should be possible to generate JSON from cards', () => {
      shallowWrapper = shallow(
        <Board {...mockProps} cards={cards} users={users} />
      );
      expect((shallowWrapper.instance() as any).exportJson()).toMatchSnapshot();
    });

    it('should export as CSV if requested', () => {
      const spy = jest.fn(() => {});
      shallowWrapper = shallow(
        <Board
          {...mockProps}
          cards={cards}
          users={users}
          setupCompleted={true}
        />
      );
      const instance = shallowWrapper.instance() as any;

      instance.handleExportCsv = spy;
      instance.handleExport('csv');
      expect(spy).toHaveBeenCalled();
    });

    it('should export as PDF if requested', () => {
      const spy = jest.fn(() => {});
      shallowWrapper = shallow(
        <Board
          {...mockProps}
          cards={cards}
          users={users}
          setupCompleted={true}
        />
      );
      const instance = shallowWrapper.instance() as any;

      instance.handleExportPdf = spy;
      instance.handleExport('pdf');
      expect(spy).toHaveBeenCalled();
    });
  });

  describe('phases', () => {
    it('should call onSwitchPhaseIndex with correct index, when next phase is called', () => {
      const spy = jest.fn();
      shallowWrapper = shallow(
        <Board {...mockProps} onSwitchPhaseIndex={spy} />
      );
      (shallowWrapper.instance() as any).handleNextPhase();
      expect(spy).toHaveBeenCalledWith(+1);
    });
    it('should call onSwitchPhaseIndex with correct index, when previous phase is called', () => {
      const spy = jest.fn();
      shallowWrapper = shallow(
        <Board {...mockProps} onSwitchPhaseIndex={spy} />
      );
      (shallowWrapper.instance() as any).handlePrevPhase();
      expect(spy).toHaveBeenCalledWith(-1);
    });
  });

  it('should show that app is loading if boardConfig is not ready yet', () => {
    shallowWrapper = shallow(
      <Board {...mockProps} boardConfig={undefined as any} />
    );
    expect(shallowWrapper.find(LoadingScreen)).toHaveLength(1);
  });

  it('should that user must be registered if setup is not completed yet', () => {
    shallowWrapper = shallow(<Board {...mockProps} setupCompleted={false} />);
    expect(shallowWrapper.find(LoadingScreen)).toHaveLength(1);
  });
});
