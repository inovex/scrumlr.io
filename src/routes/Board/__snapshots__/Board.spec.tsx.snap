// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> export should be possible to generate CSV from cards 1`] = `
"data:text/csv;charset=utf-8,sep=|
Type|Text|Author|Votes|Timestamp
positive|Text card 1|User1||2017-07-31T16:30:39.537Z
negative|Text card 2|User2||2017-07-31T16:30:39.537Z
positive|Text card 3|User1||2017-07-31T16:30:39.537Z
"
`;

exports[`<Board /> export should be possible to generate JSON from cards 1`] = `
Object {
  "cards": Array [
    Object {
      "authorref": "user1",
      "childCards": Array [
        Object {
          "authorref": "user1",
          "childCards": Array [],
          "text": "Text card 3",
          "timestamp": "2017-07-31T16:30:39.537Z",
          "type": "positive",
          "votes": -4,
        },
      ],
      "text": "Text card 1",
      "timestamp": "2017-07-31T16:30:39.537Z",
      "type": "positive",
      "votes": 13,
    },
    Object {
      "authorref": "user2",
      "childCards": Array [],
      "text": "Text card 2",
      "timestamp": "2017-07-31T16:30:39.537Z",
      "type": "negative",
      "votes": -1,
    },
  ],
  "date": "",
  "title": "Unnamed Board",
  "users": Array [
    Object {
      "id": "user1",
      "name": "User1",
    },
    Object {
      "id": "user2",
      "name": "User2",
    },
  ],
}
`;
